x/concentrated-liquidity/position.go

```diff
@@ -621,17 +610,54 @@ func (k Keeper) RemovePositionIdToLock(ctx sdk.Context, positionId, underlyingLo
 	store.Delete(lockIdKey)
 }
 
-// positionHasUnderlyingLockInState checks if a given positionId has a corresponding lock in state.
-func (k Keeper) positionHasUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, error) {
+// PositionHasActiveUnderlyingLock is a non mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, returns false.
+func (k Keeper) PositionHasActiveUnderlyingLock(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
 	// Get the lock ID for the position.
-	_, err := k.GetLockIdFromPositionId(ctx, positionId)
-	if err == nil || !errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return true, nil
+	lockId, err = k.GetLockIdFromPositionId(ctx, positionId)
+	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return false, 0, err
+	}
+	if lockIsMature {
+		return false, 0, nil
 	}
+	return true, lockId, nil
+}
+
+// positionHasActiveUnderlyingLockAndUpdate is a mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, removes the lock ID from the position ID to lock ID mapping and returns false.
+func (k Keeper) positionHasActiveUnderlyingLockAndUpdate(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
+	// Get the lock ID for the position.
+	lockId, err = k.GetLockIdFromPositionId(ctx, positionId)
 	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return false, nil
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return true, 0, err
```

Comment:
I still don't understand the `true` here. Especially given the fact that it is false in a similar branch of `PositionHasActiveUnderlyingLock`.

The complexity and similarity of both `positionHasActiveUnderlyingLockAndUpdate` and `PositionHasActiveUnderlyingLock` is the primary reason why I am also suggesting to call `PositionHasActiveUnderlyingLock` from  `positionHasActiveUnderlyingLockAndUpdate`

Reply:
Changed https://github.com/osmosis-labs/osmosis/pull/5042/commits/5821706c879fb2f2ad772a4b4da0c8333dabc45f